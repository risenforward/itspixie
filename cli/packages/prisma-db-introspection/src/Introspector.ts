import { PostgresConnector } from './connectors/PostgresConnector'
import { SdlPrinter } from './SdlPrinter'
import { ClientConfig } from 'pg'
import { PostgresConnectionDetails } from './types/common'

export class Introspector {
  connectionString: string | ClientConfig
  connector: PostgresConnector
  printer = new SdlPrinter()

  constructor(connectionString: PostgresConnectionDetails) {
    this.connectionString = connectionString
    this.connector = new PostgresConnector(connectionString)
  }

  async listSchemas(): Promise<string[]> {
    return this.connector.listSchemas()
  }

  async introspect(schemaName: string): Promise<{ numTables: number; sdl: string }> {
    const tables = await this.connector.listTables(schemaName)
    const sdl = this.printer.print(tables)

    return { numTables: tables.length, sdl }
  }
}
