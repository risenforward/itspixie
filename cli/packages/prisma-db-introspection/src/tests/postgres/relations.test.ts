import { Introspector } from '../../Introspector'
import { Client } from 'pg'

function introspect(): Promise<string> {
  return new Introspector('').introspect('DatabaseIntrospector')
}

async function testSchema(sql: string) {
  const client = new Client({ connectionString: '' })
  await client.connect()
  await client.query('DROP SCHEMA IF EXISTS DatabaseIntrospector cascade;')
  await client.query('CREATE SCHEMA DatabaseIntrospector;')
  await client.query('SET search_path TO DatabaseIntrospector;')
  await client.query(sql)

  expect(await introspect()).toMatchSnapshot()

  await client.end()
}

describe('Introspector', () => {
  test('relation with relation table', async () => {
    await testSchema(`CREATE TABLE product (
      id         serial PRIMARY KEY  -- implicit primary key constraint
    , product    text NOT NULL
    );
    
    CREATE TABLE bill (
      id       serial PRIMARY KEY
    , bill     text NOT NULL
    );
    
    CREATE TABLE bill_product (
      bill_id    int REFERENCES bill (id) ON UPDATE CASCADE ON DELETE CASCADE
    , product_id int REFERENCES product (id) ON UPDATE CASCADE
    );`)
  })

  test('relation with inline relation column', async () => {
    await testSchema(`CREATE TABLE product (
      id           serial PRIMARY KEY  -- implicit primary key constraint
    , description  text NOT NULL
    );
    
    CREATE TABLE bill (
      id         serial PRIMARY KEY
    , bill       text NOT NULL
    , product_id int REFERENCES product (id) ON UPDATE CASCADE
    );`)
  })
})
