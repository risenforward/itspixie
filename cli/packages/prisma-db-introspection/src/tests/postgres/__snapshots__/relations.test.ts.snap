// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Introspector relation with inline relation column 1`] = `
Object {
  "numTables": 2,
  "sdl": "type Bill @pgTable(name: \\"bill\\") {
  id: Int! @unique
  bill: String!
  product: Product @pgRelation(column: \\"product_id\\")
}

type Product @pgTable(name: \\"product\\") {
  id: Int! @unique
  description: String!
  bills: [Bill!]!
}
",
}
`;

exports[`Introspector relation with inline relation column NOT NULL 1`] = `
Object {
  "numTables": 2,
  "sdl": "type Bill @pgTable(name: \\"bill\\") {
  id: Int! @unique
  bill: String!
  product: Product! @pgRelation(column: \\"product_id\\")
}

type Product @pgTable(name: \\"product\\") {
  id: Int! @unique
  description: String!
  bills: [Bill!]!
}
",
}
`;

exports[`Introspector relation with relation table 1`] = `
Object {
  "numTables": 3,
  "sdl": "type Bill @pgTable(name: \\"bill\\") {
  id: Int! @unique
  bill: String!
  products: [Product!]! @pgRelationTable(table: \\"bill_product\\" name: \\"bill_product\\")
}

type Product @pgTable(name: \\"product\\") {
  id: Int! @unique
  product: String!
  bills: [Bill!]! @pgRelationTable(table: \\"bill_product\\" name: \\"bill_product\\")
}
",
}
`;

exports[`Introspector relation with relation table with extra column 1`] = `
Object {
  "numTables": 3,
  "sdl": "type Bill @pgTable(name: \\"bill\\") {
  id: Int! @unique
  bill: String!
  bill_products: [Bill_product!]!
}

type Bill_product @pgTable(name: \\"bill_product\\") {
  bill: Bill @pgRelation(column: \\"bill_id\\")
  product: Product @pgRelation(column: \\"product_id\\")
  some_other_column: String!
}

type Product @pgTable(name: \\"product\\") {
  id: Int! @unique
  product: String!
  bill_products: [Bill_product!]!
}
",
}
`;
